### ISSUE TITLE
Custom models networking (out of the box) with bokeh.client and push_session

### ISSUE THREAD

**Issue by geryatejina** [bokeh/bokeh#5296]
Hi  everyone,
as suggested by @bryevdv, I am posting here a question I've first asked in the mailing list ([here](https://groups.google.com/a/continuum.io/forum/#!mydiscussions/bokeh/boJqKJRDM8g)) and largely related to #3930.
In the code below (app.py) I am creating a custom model that simply modifies `ResetTool` to make it a "ResetView", which resets the axis ranges but retains the data selection. 

app.py:
```python
from bokeh.models import ColumnDataSource, Tool, WheelZoomTool, BoxSelectTool
from bokeh.plotting import figure
from bokeh.io import curdoc, push_session

class ResetView(Tool):
    __implementation__ = """
        ActionTool = require "models/tools/actions/action_tool"

        class ResetViewView extends ActionTool.View
          do: () ->
            @plot_view.clear_state()
            @plot_view.reset_range()

        class ResetView extends ActionTool.Model
          default_view: ResetViewView
          type: "ResetTool"
          tool_name: "Reset view"
          icon: "bk-tool-icon-reset"

        module.exports = { 
          Model: ResetView
          View: ResetViewView
        }
        """

source = ColumnDataSource(data=dict(x=[1, 2, 3, 4, 5], y=[1, 7, 9, 4, 6]))
plot = figure(x_range=(0,10), y_range=(0,10), tools=[ResetView(), WheelZoomTool(), BoxSelectTool()])
plot.circle('x', 'y', source=source)

doc = curdoc()
doc.add_root(plot)
session = push_session(doc)
session.show() 
session.loop_until_closed()  
```

This example works well if launched as a Bokeh app (`>>> bokeh serve app.py`), but fails when using `bokeh.client` and `push_session` as in:
```bash
>>> bokeh serve
```
and separately
```bash
>>> python app.py
```

It gives the following error:
```
RuntimeError: Failed to push document: KeyError("View model name 'ResetView' not found",)
```

Using the workaround proposed by @the-neworder in #3930 **does work**. However, the question is whether this code is supposed to work *as is* or if instead custom models are not intended to be supported in the "push_session" way of doing things.

As a related suggestion, I would find it very useful to have an option to the ResetTool to be able to retain the selection, something like:
```python
reset = ResetTool(reset_selection=False)
```
In fact, every time I've used Bokeh (to a great satisfaction!) I found myself wanting to reset the view but ideally retain the selection.
Thanks  a lot for any feedback and for the fantastic job with Bokeh!
Gerardo

**Comment by bryevdv** [comment]
Hi @geryatejina thanks for all the info. This is actually a dupe of #3930 (so I will close it now). I did now realize there was already an issue when I suggested you make this one, my apologies. The issue as described is that the models need to live in the server too. When you run an "native app" that just happens automatically. But with `push_session` there is nothing that automatically loads any extensions to make them available. I think the workaround is perfectly reasonable (I would not discourage it, at any rate) although it seems tedious. I am not sure how much improvement can be made, but am open to ideas, e.g. trying to pickle things to send to the server, having a config file to specify imports, etc. in #3930
